src/policy/ancestor_packages.h

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2021-2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_POLICY_ANCESTOR_PACKAGES_H
+#define BITCOIN_POLICY_ANCESTOR_PACKAGES_H
+
+#include <policy/feerate.h>
+#include <policy/packages.h>
+
+#include <vector>
+
+/** A potential BIP331 Ancestor Package, i.e. one transaction with its set of ancestors.
+ * This class does not have any knowledge of chainstate, so it cannot determine whether all
+ * unconfirmed ancestors are present. Its constructor accepts any list of transactions that
+ * IsConsistent(), linearizes them topologically, and determines whether it IsAncestorPackage().
+ * If fee and vsizes are given for each transaction, it can also linearize the transactions using
+ * the ancestor score-based mining algorithm via MiniMiner.
+ *
+ * Skip() and SkipWithDescendants() can be used to omit transactions. Txns() returns all
+ * transactions linearized, and FilteredTxns() is the same but excludes the skipped ones.
+ * GetAncestorSet() can be used to get a transaction's "subpackage," i.e. ancestor set within the
+ * package, also excluding skipped ones.
+ * */
+class AncestorPackage
+{
+    /** Whether m_txns contains a connected package in which all transactions are ancestors of the
+     * last transaction. This object is not aware of chainstate. So if m_txns only includes a
+     * grandparent and not the "connecting" parent, this will (incorrectly) determine that the
+     * grandparent is not an ancestor.
+     * */
+    bool m_is_ancestor_package{false};
```

Comment:
The name could be more verbose to reflect the notion of “connection” it lays on, e.g `m_is_ancestor_package`.